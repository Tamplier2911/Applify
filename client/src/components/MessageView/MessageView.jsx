// import "./MessageView.scss";
import React from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { deleteMessageStart } from "../../redux/messages/messages.actions";

// components
import Button from "../Button/Button";

// JS Rendering CSS
import {
  MessageViewContainer,
  MessageViewPortrait,
  MessageViewImageWrap,
  MessageViewImage,
  MessageViewContent,
  MessageViewName,
  MessageViewEmail,
  MessageViewMessage,
  MessageViewDate,
  MessageViewControlls,
  MessageViewLink
} from "./MessageViewStyles";

// component constants
import messageViewData from "../../utils/ComponentMessageViewConstants/componentMessageViewConstants";

const MessageView = ({
  createdAt,
  _id,
  name,
  email,
  message,
  from,
  lang,
  deleteMessageStart
}) => {
  let locationHref = window.location.href.split("/");
  locationHref = "".concat(locationHref[0], "//", locationHref[2], "/");
  console.log(locationHref, "location");
  // FIX IMAGE LOADING ALGORITHM
  // ISSUES IS /profile/ find way to rid of profile
  //                                    \/
  // https://applify-s.herokuapp.com/profile/api/uploads/images/users/user-5e6e618672e9151d503701ed-1584642619899.jpeg

  let photo = "";
  if (from) {
    photo = from.photo;
  } else {
    photo = "uploads/images/users/default.jpg";
  }
  console.log(photo, "photo");

  const date = new Date(createdAt).toLocaleString("en-us", {
    day: "numeric",
    month: "long",
    year: "numeric"
  });

  let image = "";
  if (process.env.NODE_ENV === "development" && photo) {
    image = `${locationHref + "api/" + photo}`;
  } else if (photo) {
    image = `${locationHref + "api/" + photo}`;
  }
  console.log(image, "image");

  const { messageViewRead, messageViewDelete } = messageViewData[lang];

  return (
    <MessageViewContainer>
      <MessageViewPortrait>
        <MessageViewImageWrap>
          <MessageViewImage alt="happy user" src={`${image}`} />
        </MessageViewImageWrap>
      </MessageViewPortrait>
      <MessageViewContent>
        <MessageViewName>{name}</MessageViewName>
        <MessageViewEmail>{email}</MessageViewEmail>
        <MessageViewMessage>
          {message.slice(0, 200) + (message[201] ? "..." : "")}
        </MessageViewMessage>
        <MessageViewDate>{date}</MessageViewDate>
      </MessageViewContent>
      <MessageViewControlls>
        <Button
          type="button"
          value={messageViewDelete}
          click={() => deleteMessageStart(_id)}
        />
        <MessageViewLink to={`/profile/messages/${_id}`}>
          {messageViewRead}
        </MessageViewLink>
      </MessageViewControlls>
    </MessageViewContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage
});

export default connect(mapStateToProps, { deleteMessageStart })(MessageView);
