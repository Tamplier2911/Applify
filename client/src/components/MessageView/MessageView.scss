/*

.message-view {
  display: grid;
  //   display: flex;
  //   flex-direction: column;
  grid-template-columns: min-content 1fr;
  grid-column-gap: 1rem;
  grid-row-gap: 1rem;
  padding: 1rem;
  box-shadow: 0rem 0rem 0.5rem #333;
  transition: box-shadow 0.3s;

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }

  @media only screen and (max-width: 520px) {
    grid-template-columns: 1fr;
  }

  &__image {
    grid-row: 1 / 3;
    align-self: center;

    @media only screen and (max-width: 520px) {
      grid-row: 1 / 2;
      justify-self: center;
    }

    &--wrapper {
      width: 10rem;
      height: 10rem;
      overflow: hidden;
      border-radius: 50%;
      box-shadow: 0rem 0rem 0.5rem var(--cl-primary);
    }

    &--img {
      width: 100%;
      height: 100%;
      object-fit: cover;
      display: block;
    }
  }

  &__body {
    display: grid;
    grid-row-gap: 0.5rem;

    &--name {
      width: 50rem;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;

      @media only screen and (max-width: 720px) {
        width: 40rem;
      }

      @media only screen and (max-width: 425px) {
        width: 30rem;
      }

      @media only screen and (max-width: 300px) {
        width: 20rem;
      }
    }
    &--email {
      width: 50rem;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;

      font-size: 1.8rem;
      font-weight: bold;

      @media only screen and (max-width: 720px) {
        width: 40rem;
      }

      @media only screen and (max-width: 425px) {
        width: 30rem;
      }

      @media only screen and (max-width: 300px) {
        width: 20rem;
      }
    }
    &--message {
    }
    &--date {
      display: grid;
      justify-items: start;
      color: var(--cl-darkest-grey);
    }
  }
  &__controlls {
    grid-column: 2 / -1;
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-column-gap: 2rem;

    @media only screen and (max-width: 520px) {
      grid-column: 1 / -1;
    }
    &--link {
      &:link,
      &:visited {
        display: grid;
        align-items: center;
        font-family: var(--font-title);
        font-size: 1.6rem;
        text-align: center;
        text-decoration: none;
        color: var(--cl-white);
        background-color: var(--cl-font);
        transition: color 0.3s, background-color 0.3s;
      }
      &:active,
      &:hover {
        background-color: var(--cl-primary);
      }
    }
  }
}

*/
