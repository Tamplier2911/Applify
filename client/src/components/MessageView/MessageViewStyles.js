import styled, { css } from "styled-components";
import { Link } from "react-router-dom";
import ScrollAnimation from "react-animate-on-scroll";

const MessageViewContentSharedStyles = css`
  width: 50rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  @media only screen and (max-width: 720px) {
    width: 40rem;
  }

  @media only screen and (max-width: 425px) {
    width: 30rem;
  }

  @media only screen and (max-width: 300px) {
    width: 20rem;
  }
`;

export const MessageViewContainer = styled(ScrollAnimation)`
  display: grid;
  grid-template-columns: min-content 1fr;
  grid-column-gap: 1rem;
  grid-row-gap: 1rem;
  padding: 1rem;
  box-shadow: 0rem 0rem 0.5rem var(--cl-darkest-grey);
  transition: box-shadow 0.3s;
  font-family: var(--font-title);

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }

  @media only screen and (max-width: 520px) {
    grid-template-columns: 1fr;
  }
`;

export const MessageViewPortrait = styled.div`
  grid-row: 1 / 3;
  align-self: center;

  @media only screen and (max-width: 520px) {
    grid-row: 1 / 2;
    justify-self: center;
  }
`;

export const MessageViewImageWrap = styled.div`
  width: 10rem;
  height: 10rem;
  overflow: hidden;
  border-radius: 50%;
  box-shadow: 0rem 0rem 0.5rem var(--cl-primary);
`;

export const MessageViewImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  display: block;
`;

export const MessageViewContent = styled.div`
  display: grid;
  grid-row-gap: 0.5rem;
`;

export const MessageViewName = styled.h3`
  ${MessageViewContentSharedStyles}
`;

export const MessageViewEmail = styled.div`
  ${MessageViewContentSharedStyles}
`;

export const MessageViewMessage = styled.div``;

export const MessageViewDate = styled.div`
  display: grid;
  justify-items: start;
  color: var(--cl-darkest-grey);
`;

export const MessageViewControlls = styled.div`
  grid-column: 2 / -1;
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-column-gap: 2rem;

  @media only screen and (max-width: 520px) {
    grid-column: 1 / -1;
  }
`;

export const MessageViewLink = styled(Link)`
  &:link,
  &:visited {
    display: grid;
    align-items: center;
    font-family: var(--font-title);
    font-size: 1.6rem;
    text-align: center;
    text-decoration: none;
    color: var(--cl-white);
    background-color: var(--cl-font);
    transition: color 0.3s, background-color 0.3s;
  }
  &:active,
  &:hover {
    background-color: var(--cl-primary);
  }
`;
