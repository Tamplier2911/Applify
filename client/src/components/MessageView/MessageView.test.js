import React from "react";

// component
import MessageView from "./MessageView";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
  checkProps,
} from "../../tests/testUtils";

describe("<MessageView />", () => {
  const initialState = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
  };
  const store = storeFactory(initialState);

  const message = {
    createdAt: "10-05-2020",
    _id: "sdk321",
    name: "Breadberry",
    email: "ex@mple.com",
    message: "Well, hello!",
    form: {},
  };

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<MessageView store={store} {...message} />);
  });

  it("renders without an error", () => {
    const messageView = findByTestAttr(wrapper, "message-view");
    expect(messageView.length).toBe(1);
  });

  it("renders children - message content", () => {
    const messageViewContent = findByTestAttr(wrapper, "message-view-content");
    expect(messageViewContent.length).toBe(1);
  });

  it("renders children - message controls", () => {
    const messageViewControls = findByTestAttr(
      wrapper,
      "message-view-controls"
    );
    expect(messageViewControls.length).toBe(1);
  });

  it("recieves expected props", () => {
    const expectedProps = {
      createdAt: "10-05-2020",
      _id: "sdk321",
      name: "Breadberry",
      email: "ex@mple.com",
      message: "Well, hello!",
      form: {},
    };
    const err = checkProps(MessageView, expectedProps, "props");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const messageView = findByTestAttr(wrapper, "message-view-content");
    expect(messageView).toMatchSnapshot();
  });
});
