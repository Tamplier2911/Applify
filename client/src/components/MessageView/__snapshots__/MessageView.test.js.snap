// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageView /> matches snapshot 1`] = `
<styled.div
  data-test="message-view-content"
>
  <styled.h3>
    Breadberry
  </styled.h3>
  <styled.div>
    ex@mple.com
  </styled.div>
  <styled.div>
    Well, hello!
  </styled.div>
  <styled.div>
    October 5, 2020
  </styled.div>
</styled.div>
`;
